core/src/main/kotlin/transfer/Uploader.kt

```diff
@@ -0,0 +1,138 @@
+package com.juul.mcumgr.transfer
+
+import com.juul.mcumgr.McuMgrResult
+import com.juul.mcumgr.message.Format
+import com.juul.mcumgr.onErrorOrFailure
+import com.juul.mcumgr.onSuccess
+import java.lang.IllegalArgumentException
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.launch
+
+internal const val TRUNCATED_HASH_LEN = 3
+private const val RETRIES = 5
+
+interface Uploader {
+
+    val mtu: Int
+    val format: Format
+
+    suspend fun write(data: ByteArray, offset: Int, length: Int): McuMgrResult<Response>
+
+    data class Event(val offset: Int, val length: Int)
+    data class Response(val offset: Int)
+}
+
+@Throws
+suspend fun Uploader.flow(
```

Comment:
I don't quite follow why this is a `Flow`?
Perhaps I'm just having trouble understanding what this function is doing.

Is it grabbing chunks of data that will ultimately be uploaded?

Reply:
Resolving to move discussion to other comment